soh/soh/Enhancements/accessible-actors/SfxExtractor.cpp

```diff
@@ -0,0 +1,251 @@
+    void* mem = NULL;
+    size_t size = 0;
+
+    if (!drwav_init_memory_write(&wav, &mem, &size, &format, NULL))
+        throw std::runtime_error("SFX Extractor: Unable to initialize wave writer.");
+    int16_t chunk[64];
+    int16_t* mark = tempBuffer + startOfInput;
+    size_t samplesLeft = endOfInput - startOfInput;
+    while (samplesLeft > 0) {
+        size_t thisChunk = std::min<size_t>(64, samplesLeft);
+        ma_channel_converter_process_pcm_frames(&converter, chunk, mark, thisChunk / 2);
+        drwav_write_pcm_frames(&wav, thisChunk / 2, chunk);
+        samplesLeft -= thisChunk;
+        mark += thisChunk;
+    }
+    drwav_uninit(&wav);
+    archive->AddFile(fileName.c_str(), (uintptr_t)mem, size);
+
+    drwav_free(mem, NULL);
+}
+void SfxExtractor::setup() {
+    try {
+
+        SpeechSynthesizer::Instance->Speak("Sfx extraction speedrun initiated. Please wait. This will take a few minutes.", GetLanguageCode());
+        // Kill the audio thread so we can take control.
+        captureThreadState = CT_WAITING;
+        OTRAudio_InstallSfxCaptureThread();
+        // Make sure we're starting from a clean slate.
+        std::string sohAccessibilityPath = LUS::Context::GetPathRelativeToAppDirectory("accessibility.otr");
+        if (std::filesystem::exists(sohAccessibilityPath)) {
+            currentStep = STEP_ERROR_OTR;
+            return;
+        }
+        tempStorage.resize((SFX_EXTRACTION_BUFFER_SIZE + (SFX_EXTRACTION_ONE_FRAME * 3)) * 2,
+                           0); // Over-allocated just a tad because otherwise we'll overrun if the last frame is short.
+        tempBuffer = tempStorage.data();
+
+        // Build are master sfx extraction todo list.
+        for (int i = 0; i < sfxCount; i++)
+            sfxToRip.push(sfxTable[i]);
+
+        currentStep = STEP_MAIN;
+        for (int i = 1; i < 10; i++)
+            progressMilestones[i - 1] = sfxToRip.size() - ((int)ceil(sfxToRip.size() * (i / 10.0f)));
+        archive = LUS::Archive::CreateArchive("accessibility.otr", sfxToRip.size());
+
+    } catch (...) { currentStep = STEP_ERROR; }
+}
+void SfxExtractor::ripNextSfx() {
+    {
+        auto lock = OTRAudio_Lock();
+        if (captureThreadState == CT_READY || captureThreadState == CT_PRIMING)
+            return; // Keep going.
+    }
+    // Was the last sfx a loop? If so then we need to stop it, and then we need to run audio out to nowhere for as long
+    // as it takes to get back to a blank slate.
+    if (currentSfx != -1) {
+        Audio_StopSfxByPos(&D_801333D4);
+        captureThreadState = CT_PRIMING;
+        currentSfx = -1;
+
+        return;
+    }
+    if (sfxToRip.empty()) {
+        currentStep = STEP_FINISHED; // Caught 'em all!
+        return;
+    }
+
+    currentSfx = sfxToRip.front();
+    sfxToRip.pop();
+    startOfInput = 0;
+    endOfInput = 0;
+    Audio_PlaySoundGeneral(currentSfx, &D_801333D4, 4, &D_801333E0, &D_801333E0, &D_801333E8);
```

Comment:
i feel like there has to be a faster way to extract all the sfx into a format that miniaudio is happy with than looping through every sfx and recording it. ideally something we could integrate into the `oot.otr` generation process.

i'm sure i'm not the first (and won't be the last) person to think this. i'd love to hear the reasoning behind this decision and any rejected alternative methods attempted.

Reply:
Allow me to say this: I wish that getting my hands on the sound effects was easier as well. The sound system in this game is utter voodoo.

Basically, what you have is a collection of about 400 raw audio samples. They're tiny; most are less than a second long. We can decode these directly (and in fact we do in some cases), but by themselves they are extremely primitive.
From this small library, well over an hour of actual sound effects are created.

The creation itself is done via a full-blown programming language. The sequence files are some form of assembly-like bytecode generated from a script: think of it as Midi++++++.

I did speak to Sauren (the author of Seq64) about this.
Extracting them outside of the game would probably require integrating the majority of Seq64. Or studying the assembly and building a new player. Or copying the entire audio infrastructure from the game into a separate tool. This would be a herculean task even if audio were better understood/ documented.

That said, the extraction is not realtime recording. It runs in about two minutes, hence the "speedrun" reference you asked about. It plays the sound, then runs audio as fast as possible until the entire sound effect is obtained.

I think the biggest disadvantage of using these extracted assets is that sounds that are supposed to loop don't do so perfectly as they do when played normally. We've largely avoided looping sounds so far in order to avoid this.
It's certainly 
not perfect, but it works quite well.

Reply:
Hi! First of all thank you so much for working on accessibility for this amazing game I always wanted to play but I couldn't! :)

But let me introduce myself a bit. :)

I am totally blind (sightless) drummer, singer, audio engineer, voice
actor and accessibility consultant. I am working with game developers
and modders to improve accessibility for the blind players.

I was working on accessibility for games like:

-Skullgirls - First fully blind accessible 2d fighter.
-Hades - First ever fully playable roguelike game for the blind due to
the mods created:
https://caniplaythat.com/2021/05/17/hades-accessibility-mods-for-blind-visually-impaired-players/

-Slay The Spire - First fully accessible roguelike deck-building game
thanks to it's modding system (the devs rejected working on the
accessibility so We, blind gamers decided to create the mods):
https://bradjrenshaw.github.io/say-the-spire

Lastly, I was working with French developer COWCATGames on BROK: The InvestiGator -
First and only one fully blind accessible point-and-click adventure
game with beat em up elements:
https://www.brokgame.com/accessibility.php

I've tested the latest build and I am liking it so far even if I can't understand all of the extra audio cues. That'd be good if the game can have some kind of menu with the audio glossary etc.

Another issue is these sounds are too loud and too aggressive a bit. Is there a way to use the standard .wav / other format sounds to use? If yes I can think about something and try to make the full sound design for these.

Reply:
Hi,

Thank you very much for your message!

Yes indeed! Finding appropriate sounds for everything has been a huge challenge, and what we do find is rarely ideal.

As per the code you were looking at, all we are doing at the moment is collecting all of the sounds from the game itself and outputting them to wav files,
so we can play them via the custom engine. So yes, the custom engine supports playing external .wav files; in fact that's literally the only thing it does.
It could play flac and mp3 as well, and ogg/vorbis support would also be easy to include.

Most objects in the game are currently represented by just taking the existing sounds and manipulating them in some way (pitching them, playing isolated
sections of them, etc).

We would absolutely love to have your assistance with sound design if that's something you're up for.

I will warn you however, the current list of actor sounds is significant and ever-growing. Have a look through AccessibleActorList.cpp to get an idea of
what the current scale is. We're still early in the game so this number is expected to grow considerably.

As for the other stuff, we definitely have to figure out how to add menus and such. It would be amazing if we could add "learn game sounds" and other things
to the actual pause menu, but at this point neither of us have managed to dive that deep into the game's internals. I'm sure we'll get there in due course.

Kind regards,

Jordan.

Reply:
Hi,
Great! Will definitely contribute to this as I'll do everything to play this masterpiece and with your help I can see it is possible! :)

Let me know if you're on DIscord as I'll have some more questions probably and this can be faster way to contact. If not it is OK. :)

Reply:
Yep. I'm caturria on Discord.